spam/spam_data_augmentation_tutorial.py

```diff
@@ -0,0 +1,301 @@
+#
+# We start with a simple transformation function that changes a random character in the text to simulate a typo.
+
+# %%
+import string
+from snorkel.augmentation.tf import transformation_function
+
+
+@transformation_function()
+def change_character(x):
+    idx = np.random.choice(range(len(x.text) - 1))
+    char = np.random.choice(list(string.ascii_lowercase))
+    x.text = x.text[:idx] + char + x.text[idx + 1 :]
+    return x
+
+
+# %% [markdown]
+# ### Adding `pre` mappers.
+# Some TFs rely on fields that aren't present in the raw data, but can be derived from it.
+# We can enrich our data (providing more fields for the TFs to refer to) using map functions specified in the `pre` field of the transformation_function decorator (similar to `preprocessor` used for Labeling Functions).
+#
+# For example, we can use the fantastic NLP tool [spaCy](https://spacy.io/) to add lemmas, part-of-speech (pos) tags, etc. to each token.
+# Snorkel provides a prebuilt preprocessor for spaCy called `SpacyPreprocessor` which adds a new field to the
+# data point containing a [spaCy `Doc` object](https://spacy.io/api/doc).
+# For more info, see the [`SpacyPreprocessor` documentation](https://snorkel.readthedocs.io/en/master/source/snorkel.labeling.preprocess.html#snorkel.labeling.preprocess.nlp.SpacyPreprocessor).
+#
+
+# %%
+# Download the spaCy english model
+# If you see an error in the next cell, restart the kernel
+# ! python -m spacy download en_core_web_sm
+
+# %%
+from snorkel.labeling.preprocess.nlp import SpacyPreprocessor
+
+# The SpacyPreprocessor parses the text in text_field and
+# stores the new enriched representation in doc_field
+spacy = SpacyPreprocessor(text_field="text", doc_field="doc", memoize=True)
+
+# %%
+import numpy as np
+
+# TFs for replacing a random named entity with a different entity of the same type.
+@transformation_function(pre=[spacy])
+def change_person(x):
+    persons = [str(ent) for ent in x.doc.ents if ent.label_ == "PERSON"]
+    if persons:
+        to_replace = np.random.choice(persons)
+        replacement = np.random.choice(["Bob", "Alice"])
+        x.text = x.text.replace(to_replace, replacement)
+        return x
+
+
+@transformation_function(pre=[spacy])
+def change_date(x):
+    dates = [str(ent) for ent in x.doc.ents if ent.label_ == "DATE"]
+    if dates:
+        to_replace = np.random.choice(dates)
+        replacement = np.random.choice(["31st December", "01/03/99"])
+        x.text = x.text.replace(to_replace, replacement)
+        return x
+
+
+# Drop the last sentence of a multi-sentence comment, as this shouldn't change it's spam / ham nature.
+@transformation_function(pre=[spacy])
+def drop_last_sentence(x):
+    sentences = [str(span) for span in x.doc.sents]
+    if len(sentences) > 1:
+        x.text = ". ".join(sentences[:-1])
```

Comment:
is the string repr of a spacy sentence just the original text?

Reply:
Yes, but replaced with .text since it's documented.